workers/unity/Packages/com.improbable.gdk.gameobjectcreation/GameObjectInitializationSystem.cs

```diff
@@ -0,0 +1,109 @@
+    [UpdateInGroup(typeof(SpatialOSReceiveGroup.EntityInitialisationGroup))]
+    [DisableAutoCreation]
+    internal class GameObjectInitializationSystem : ComponentSystem
+    {
+        private struct AddedEntitiesData
+        {
+            public readonly int Length;
+            public EntityArray Entities;
+            [ReadOnly] public ComponentDataArray<SpatialEntityId> SpatialEntityIds;
+            [ReadOnly] public ComponentDataArray<NewlyAddedSpatialOSEntity> NewlyCreatedEntities;
+        }
+
+        private struct RemovedEntitiesData
+        {
+            public readonly int Length;
+            public EntityArray Entities;
+            [ReadOnly] public ComponentDataArray<GameObjectReferenceHandle> GameObjectReferenceHandles;
+            public SubtractiveComponent<GameObjectReference> NoGameObjectReference;
+        }
+
+        [Inject] private AddedEntitiesData addedEntitiesData;
+        [Inject] private RemovedEntitiesData removedEntitiesData;
+
+        [Inject] private EntityGameObjectLinkerSystem LinkerSystem;
+
+        private WorkerSystem worker;
+        private ViewCommandBuffer viewCommandBuffer;
+        private readonly Dictionary<Entity, GameObject> entityGameObjectCache = new Dictionary<Entity, GameObject>();
+        private readonly IEntityGameObjectCreator gameObjectCreator;
+
+        public GameObjectInitializationSystem(IEntityGameObjectCreator gameObjectCreator)
+        {
+            this.gameObjectCreator = gameObjectCreator;
+        }
+
+        protected override void OnCreateManager(int capacity)
+        {
+            base.OnCreateManager(capacity);
+
+            worker = World.GetExistingManager<WorkerSystem>();
+            viewCommandBuffer = new ViewCommandBuffer(EntityManager, worker.LogDispatcher);
+        }
+
+        protected override void OnUpdate()
+        {
+            for (var i = 0; i < addedEntitiesData.Length; i++)
+            {
+                var entity = addedEntitiesData.Entities[i];
+                var spatialEntityId = addedEntitiesData.SpatialEntityIds[i].EntityId;
+
+                var gameObject = gameObjectCreator.CreateGameObjectForEntityAdded(
+                    new SpatialOSEntity(entity, EntityManager), worker);
+
+                if (gameObject == null)
+                {
+                    continue;
+                }
+
+                LinkerSystem.Linker.LinkGameObjectToEntity(gameObject, entity, spatialEntityId, PostUpdateCommands, viewCommandBuffer);
+                entityGameObjectCache.Add(entity, gameObject);
+            }
+
+            for (var i = 0; i < removedEntitiesData.Length; i++)
+            {
+                var entity = removedEntitiesData.Entities[i];
+
+                if (!entityGameObjectCache.TryGetValue(entity, out var gameObject))
+                {
+                    // Entity without linked GameObject removed
+                    gameObjectCreator.OnEntityRemoved(new SpatialOSEntity(entity, EntityManager),
```

Comment:
Nit: this can be moved outside of the if/else loop if you set `gameObject` to be null in the "try get failed" branch